---
title: '20210702'
date: 2021-07-02 23:07:90
category: TIL
thumbnail: { thumbnailSrc }
draft: false
---

## 스캡프로젝트

- redux, asyncStorage, fs 수정

  - 기존코드는 내부적으로 try-catch를 썼더니 redux에서 error 분기점을 하기 어려운 문제가 있어 상위 try-catch로 넘기는 방향으로 수정함
  - 그리고 기본폴더 예외케이스를 위해 IStoredFolder 타입에 filePath를 추가했다.

  ```tsx
  export interface IStoredFolder {
    id: string
    folderName: string
    borderColor: string
    filePath: string
  }
  ```

  - redux에서 변경된 코드. try-catch를 redux 내에서 처리하게 했다.
    - 문제는 파일시스템, asyncStorage 둘 중 하나가 에러를 생성했을 때 롤백 기능과 같은 에러 처리가 어렵다는 점이다.
    - 추후에 async thunk를 파일시스템, asyncStorage 각각 만들어 파이프라인으로 연결해서 처리하는 작업을 진행해야 한다.
    - 특히 파일시스템은 통과가 되었는데 asyncStorage에서 문제가 발생하면 asyncStorage에 다시 시도를 한다든지, 몇번의 시도가 실패하면 파일시스템을 복구하는 등의 추가 작업이 필요하다.
    - 또한 엣지케이스인 유저가 내장메모리를 직접 건드렸을 경우를 고려해 앱을 매번 열때마다 스캡폴더에 접근해서 asyncStorage와 동기화 처리를 해야 할 것이다.
    - 그리고 곧바로 에러를 throw하면 rejected로 다 넘어가는 줄 알았으나 일부 에러는 그렇지 않은 경우가 있어 thunkAPI로 직접 reject시켜야 한다는 걸 알게되었다.
    - thunkAPI에 full 시키는 기능도 있길래 extraReducers를 대체하는 api로 생각해서 extraReducers를 다 삭제해봤다. 결과는 놉! redux state에 제대로 반영이 되지 않아서 이부분은 다시 살렸다.

  ```tsx
  const increment = createAsyncThunk(
    'increment',
    async (newUserFolder: IStoredFolder, thunkAPI) => {
      const { entries, loading } = thunkAPI.getState()
        .folders as IUserFolderState
      try {
        const newEntries = [...entries, newUserFolder]
        await FS.createFolder(newUserFolder.folderName)
        await storage.setUserFolders(newEntries)
        return { entries: newEntries, loading }
      } catch (error) {
        console.error(error)
        return thunkAPI.rejectWithValue({ entries, loading: LOADING.FAILED })
      }
    }
  )
  ```
